File: src/main/java/net/yakclient/mixin/internal/methodadapter/tree/TreeMixinPatternMatcher.java
package net.yakclient.mixin.internal.methodadapter.tree;

import net.yakclient.mixin.internal.instruction.Instruction;
import net.yakclient.mixin.internal.instruction.tree.TreeInsnExecutor;
import net.yakclient.mixin.internal.instruction.ASMInstruction;
import net.yakclient.mixin.internal.methodadapter.MixinPatternMatcher;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.InsnList;

public abstract class TreeMixinPatternMatcher implements MixinPatternMatcher {
    private final Instruction mixinInsn;

    public TreeMixinPatternMatcher(Instruction instructions) {
        this.mixinInsn = instructions;
    }

    public void insertInsn(AbstractInsnNode index, InsnList list) {
        new TreeInsnExecutor(list, index, this.mixinInsn).execute();
    }


    public abstract void transform(InsnList insn);
}

File: base/src/main/kotlin/net/yakclient/mixins/base/internal/methodadapter/tree/TreeMixinPatternMatcher.kt
package net.yakclient.mixins.base.internal.methodadapter.tree

import net.yakclient.mixins.base.internal.instruction.Instruction
import net.yakclient.mixins.base.internal.instruction.tree.InitialTreeInsnExecutor
import net.yakclient.mixins.base.internal.instruction.tree.PlacedTreeInsnExecutor
import net.yakclient.mixins.base.internal.methodadapter.MixinPatternMatcher
import org.objectweb.asm.tree.AbstractInsnNode
import org.objectweb.asm.tree.InsnList

abstract class TreeMixinPatternMatcher(private val mixinInsn: Instruction?) : MixinPatternMatcher {
    fun insertInsn(index: AbstractInsnNode?, list: InsnList) {
        if (index != null) PlacedTreeInsnExecutor(list, index, mixinInsn!!).execute()
    }

    fun insertInsn(list: InsnList?) {
        InitialTreeInsnExecutor(list!!, mixinInsn!!).execute()
    }

    abstract fun transform(insn: InsnList)
}